<a href="https://colab.research.google.com/github/13RENDA/GE-Vernova-Capstone/blob/main/Model_training.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Training
#### Author: Kyle

In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import *
from sklearn.metrics import r2_score
from sklearn.preprocessing import StandardScaler
import warnings
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor, AdaBoostRegressor
from xgboost import XGBRegressor
from lightgbm import LGBMRegressor
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

from google.colab import drive
drive.mount('/content/drive')

df=pd.read_csv("/content/drive/MyDrive/Broward_OpenWeather.csv")
df_1=pd.read_excel("/content/drive/MyDrive/Capstone GE/Data_2.7/Broward.xlsx")

In [ ]:
y = df_set['% Power Outage']
x = df_set.drop('% Power Outage',axis=1)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

warnings.filterwarnings('ignore')

sc = StandardScaler()
x_train_scaled = sc.fit_transform(x_train)
x_test_scaled  = sc.transform(x_test)


In [ ]:
models = {
    "Linear Regression": LinearRegression(),
    "Ridge Regression": Ridge(),
    "Lasso Regression": Lasso(),
    "ElasticNet Regression": ElasticNet(),
    "SVR": SVR(),
    "Decision Tree Regression": DecisionTreeRegressor(),
    "Random Forest Regression": RandomForestRegressor(),
    "Gradient Boosting Regression": GradientBoostingRegressor(),
    "AdaBoost Regression": AdaBoostRegressor(),
    "XGBoost Regression": XGBRegressor(),
    "LightGBM Regression": LGBMRegressor()
}

In [ ]:
for name, model in models.items():
    model.fit(x_train_scaled, y_train)
    predictions = model.predict(x_test_scaled)
    r2 = r2_score(y_test, predictions)
#     mse = mean_squared_error(y_test, predictions)
#     mae = mean_absolute_error(y_test, predictions)
    print(f"{name} Performance:")
    print("R2 Score:", r2)
#     print("Mean Squared Error:", mse)
#     print("Mean Absolute Error:", mae)
    print()

In [ ]:
corr_matrix = df_set.corr(method='spearman')

plt.figure(figsize=(20, 10))
sns.heatmap(corr_matrix, annot=True, cmap='RdYlBu', vmin=-1, vmax=1, annot_kws={"size": 6, "ha": 'center'}, fmt='.2f')
plt.title('correlation')

plt.xticks(rotation=90)
plt.yticks(rotation=0)

In [ ]:
import seaborn as sns

plt.figure(figsize=(16,10 ))
sns.boxplot(data=df_set)
plt.title('Boxplot of Features')
plt.xticks(rotation=45)
plt.show()

## Training
#### Author: Sreeram

In [ ]:
# Linear Regression
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

df.dropna(inplace=True)

X = df.drop(columns=['% Power Outage','RecordDateTime'])  # Features
y = df['% Power Outage']  # Target variable

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Mean Squared Error:", mse)
print("R^2 Score:", r2)

In [ ]:
# Random Forest

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

df.dropna(inplace=True)

X = df.drop(columns=['% Power Outage','RecordDateTime'])
y = df['% Power Outage']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

rf_model = RandomForestRegressor(random_state=42)

rf_model.fit(X_train, y_train)

y_pred = rf_model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Mean Squared Error:", mse)
print("R^2 Score:", r2)

In [ ]:
# Scaled Random Forest

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

model = RandomForestRegressor(random_state=42)
model.fit(X_train_scaled, y_train)

y_pred = model.predict(X_test_scaled)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R^2 Score: {r2}")

In [ ]:
df1.dropna(inplace=True)

X = df1.drop(columns=['% Power Outage','RecordDateTime'])  # Features
y = df1['% Power Outage']  # Target variable

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Mean Squared Error:", mse)
print("R^2 Score:", r2)

In [ ]:
# Scaled Random Forest

from sklearn.preprocessing import StandardScaler

df1.dropna(inplace=True)
X = df1.drop(columns=['% Power Outage','RecordDateTime'])  # Features
y = df1['% Power Outage']  # Target variable

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
rf_model = RandomForestRegressor(random_state=42)
model.fit(X_train_scaled, y_train)
y_pred = model.predict(X_test_scaled)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R^2 Score: {r2}")

In [ ]:
df1 = pd.read_excel('SMOTE_Broward.xlsx')

#combined_df = pd.concat([df1, df2, df3], ignore_index=True)
combined_df = df1
combined_df.drop(columns=['Classification', 'RecordDateTime'], inplace=True)
combined_df.dropna(inplace=True)

X = combined_df.drop(columns=['% Power Outage'])
y = combined_df['% Power Outage']

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, shuffle = False, random_state=42)

from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(random_state=42)
model.fit(X_train, y_train)

from sklearn.metrics import mean_squared_error, r2_score
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R^2 Score: {r2}")

In [ ]:
test_data = pd.read_excel('/content/Orange.xlsx')
test_data.drop(columns=['Classification', 'RecordDateTime'], inplace=True)

X_test_new = test_data.drop(columns=['% Power Outage'])
y_test_new = test_data['% Power Outage']

y_pred_new = model.predict(X_test_new)

mse_new = mean_squared_error(y_test_new, y_pred_new)
r2_new = r2_score(y_test_new, y_pred_new)

print(f"Mean Squared Error on new data: {mse_new}")
print(f"R^2 Score on new data: {r2_new}")

In [ ]:
results_df = pd.DataFrame({'Actual': y_test_new, 'Predicted': y_pred_new})
print(results_df)

### Support Vector Machine (Regressor)

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import StandardScaler

combined_df = pd.concat([df1, df2, df3], ignore_index=True)
combined_df.drop(columns=['Classification', 'RecordDateTime'], inplace=True)
combined_df.dropna(inplace=True)

X = combined_df.drop(columns=['% Power Outage'])
y = combined_df['% Power Outage']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

model = SVR(kernel='rbf', C=1.0, epsilon=0.1)
model.fit(X_train_scaled, y_train)

y_pred = model.predict(X_test_scaled)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R^2 Score: {r2}")

Mean Squared Error: 0.014360276698856495
R^2 Score: -0.2710956151606079


In [ ]:
test_data = pd.read_excel('/content/Orange.xlsx')
test_data.drop(columns=['Classification', 'RecordDateTime'], inplace=True)

X_test_new = test_data.drop(columns=['% Power Outage'])
X_test_scaled_new = scaler.transform(X_test_new)
y_pred_new = model.predict(X_test_scaled_new)

mse_new = mean_squared_error(y_test_new, y_pred_new)
r2_new = r2_score(y_test_new, y_pred_new)

print(f"Mean Squared Error on new data: {mse_new}")
print(f"R^2 Score on new data: {r2_new}")

Mean Squared Error on new data: 1.1166762425684584
R^2 Score on new data: 0.005381497553619385


In [ ]:
results_df = pd.DataFrame({'Actual': y_test_new, 'Predicted': y_pred_new})
print(results_df)

         Actual  Predicted
0      0.000987   0.100514
1      0.000658   0.102138
2      0.000329   0.102244
3      0.000000   0.095044
4      0.016954   0.099930
...         ...        ...
26248  0.001239   0.092607
26249  0.002013   0.091507
26250  0.006193   0.087877
26251  0.001239   0.083670
26252  0.000310   0.094037

[26253 rows x 2 columns]


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import RobustScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

combined_df = pd.concat([df1, df2, df3], ignore_index=True)
combined_df.drop(columns=['Classification', 'RecordDateTime'], inplace=True)
combined_df.dropna(inplace=True)

X = combined_df.drop(columns=['% Power Outage'])
y = combined_df['% Power Outage']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

numeric_features = X.select_dtypes(include=['int', 'float']).columns
categorical_features = X.select_dtypes(include=['object']).columns

numeric_transformer = Pipeline(steps=[('scaler', RobustScaler())])
categorical_transformer = Pipeline(steps=[('onehot', OneHotEncoder(handle_unknown='ignore'))])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)])

model = Pipeline(steps=[('preprocessor', preprocessor),
                        ('regressor', SVR(kernel='rbf', C=1.0, epsilon=0.1))])

model.fit(X_train, y_train)

y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R^2 Score: {r2}")

Mean Squared Error: 0.014726502476881046
R^2 Score: -0.3035119808315325


In [ ]:
import pandas as pd
from sklearn.metrics import mean_squared_error, r2_score

test_data = pd.read_excel('/content/Orange.xlsx')

test_data.drop(columns=['Classification', 'RecordDateTime'], inplace=True)

X_test_new = test_data

y_pred_new = model.predict(X_test_new)

mse_new = mean_squared_error(y_test_new, y_pred_new)
r2_new = r2_score(y_test_new, y_pred_new)

print(f"Mean Squared Error on new data: {mse_new}")
print(f"R^2 Score on new data: {r2_new}")

Mean Squared Error on new data: 1.1174156519856417
R^2 Score on new data: 0.004722908914246315


### Result

| Model                                        | Mean Squared Error (Val Data) | R^2 Score (Val Data) | Mean Squared Error (Test Data) | R^2 Score (Test Data) |
|----------------------------------------------|---------------------|------------|-------------------------------|-----------------------|
| Random Forest Regressor                      | 0.008636774718338897| 0.235518   | 1.1026947770552067           | 0.017834725947563523  |
| Random Forest Regressor with StandardScalar  | 0.008638702123532231| 0.235347   | 1.1032414289652068           | 0.01734782555212455   |
| Gradient Boosting Regressor                  | 0.008367395809310969| 0.259362   | 1.0850713509165413           | 0.03353183227598011   |
| Support Vector Regressor                     | 0.014360276698856495| -0.271096  | 1.1166762425684584           | 0.005381497553619385  |
| Support Vector Regressor with one-hot encoding| 0.014726502476881046| -0.303512  | 1.1174156519856417           | 0.004722908914246315  |
